Fix CLI tool print entire freelist. Add option to dump leaf element keys/value bytes.

jpbetz opened the pull request:
This includes two small fixes to the bolt CLI that we've needed a couple times in the past--

- Improve `bolt page` to print an accurate freelist.  Won't be relevant for dbs that don't persist the freelist, but still an improvement to have the command print the most accurate information available.
- Introduce `bolt page-item` to print a page item key and value. Had been handy for digging into data values at the bbolt level.

heyitsanthony commented:
> like switching over to cobra

I don't think there should be dependencies in bbolt aside from the standard libraries since it was deliberately avoided in boltdb.

heyitsanthony commented on db_test.go:
```diff
@@ -450,7 +450,7 @@ func TestDB_Open_ReadOnly(t *testing.T) {
 	if err := readOnlyDB.View(func(tx *bolt.Tx) error {
 		value := tx.Bucket([]byte("widgets")).Get([]byte("foo"))
-		if bytes.Compare(value, []byte("bar")) != 0 {
+		if !bytes.Equal(value, []byte("bar")) {
```
already picked this up in e39821f3defb7f31a52ff48f5e1b01433dd713a3

jpbetz replied:
Great, I'll rebase.

heyitsanthony commented on cmd/bolt/main.go:
```diff
@@ -593,12 +645,21 @@ func (cmd *PageCommand) PrintFreelist(w io.Writer, buf []byte) error {
+	if p.count == 0xFFFF {
+		idx = 1
+		count = int(((*[maxAllocSize]pgid)(unsafe.Pointer(&p.ptr)))[0])
+		fmt.Fprintf(w, "Item Count of 65535 (max uint16) indicates a freelist overflow. Actual Item Count: %d\n", count)
```
sort of wordy/obscure, maybe only have `Item Count:` but determine the count depending on whether it's overflow or not? if the physical dump of the page structure is needed it should probably be handled separate from this

heyitsanthony commented on cmd/bolt/main.go:
```diff
@@ -653,7 +714,14 @@ func (cmd *PageCommand) PrintPage(w io.Writer, r io.ReaderAt, pageID int, pageSi
 func (cmd *PageCommand) Usage() string {
 	return strings.TrimLeft(`
-usage: bolt page -page PATH pageid [pageid...]
+usage: bolt page PATH pageid [pageid...]
```
`DumpCommand`'s Usage looks similarly broken, might as well fix that too...

jpbetz replied:
Sure enough. Fixed.

heyitsanthony commented on cmd/bolt/main.go:
```diff
@@ -562,6 +583,37 @@ func (cmd *PageCommand) PrintLeaf(w io.Writer, buf []byte) error {
+	p := (*page)(unsafe.Pointer(&pageBytes[0]))
+
+	if index >= p.count {
+		return nil, fmt.Errorf("Expected item index >= 0 and < %d, but got %d.", p.count, index)
```
```go
fmt.Errorf("leafPageElement: expected item index less than %d, got %d", p.count, index)
```

to closer match the other `expected` errors and since index and p.count are unsigned

heyitsanthony commented on cmd/bolt/main.go:
```diff
@@ -487,16 +500,24 @@ func (cmd *PageCommand) Run(args ...string) error {
-		fmt.Fprintf(cmd.Stdout, "Page ID:    %d\n", p.id)
-		fmt.Fprintf(cmd.Stdout, "Page Type:  %s\n", p.Type())
-		fmt.Fprintf(cmd.Stdout, "Total Size: %d bytes\n", len(buf))
+		if options.itemKey == -1 && options.itemValue == -1 {
```
judging by the special-casing and that `pages` is described as `print one or more pages in human readable format`, this might be better as a separate command `dump-item [-key|-value] PATH pageid idx` instead of overloading the pages command

jpbetz replied:
Humm, yeah. I had been thinking about splitting this out. I'll do it now.

xiang90 commented:
@jpbetz 

it seems like the bin file is also checked in. remove it?

jpbetz commented:
Oops, I'll remove the bin file and check the .gitignore settings.

xiang90 commented on cmd/bolt/main.go:
```diff
@@ -416,9 +419,137 @@ func (cmd *DumpCommand) PrintPage(w io.Writer, r io.ReaderAt, pageID int, pageSi
+		return cmd.PrintLeafItemKey(cmd.Stdout, buf, uint16(itemID))
+	case options.value:
+		return cmd.PrintLeafItemValue(cmd.Stdout, buf, uint16(itemID))
+	default:
```
move this to line 465, close to other validation?

xiang90 commented on cmd/bolt/main.go:
```diff
@@ -416,9 +419,137 @@ func (cmd *DumpCommand) PrintPage(w io.Writer, r io.ReaderAt, pageID int, pageSi
+		return ErrUsage
+	}
+
+	if options.key && options.value {
```
why not allow both to be set?

jpbetz replied:
That sounds reasonable. Maybe we should do something more like etcd where by default we print key and value, and provide for `--key-only` and `--value-only` flags.

xiang90 replied:
right. i hope user can use the command without specifying any flag by default. right now they have to say --key or --value. If no option is provided, it will fail. not a good UX in my opinion.

xiang90 commented:
The feature looks good in general. Let us clean up the issues mentioned in the review and get this merged. Thank you!

heyitsanthony commented on cmd/bolt/main.go:
```diff
@@ -416,9 +419,155 @@ func (cmd *DumpCommand) PrintPage(w io.Writer, r io.ReaderAt, pageID int, pageSi
+		return err
+	}
+
+	if !options.valueOnly {
```
```go
if !options.valuesOnly { ... }
if !options.keyOnly { ... }
```
so the output with no flags exactly matches the physical layout since this is a raw dump.

etcdctl's default output mode inserts a newline and in retrospect I think it makes the output really difficult to parse, but it's too late now. `printf("%q\n%q\n",key,val)` seems like the best/easiest human-readable way to print binary data, if that's worth pursuing in the future.

jpbetz replied:
How about we do something like this:

```
bolt page-item <dbfile> <pageid> <itemid> 
"<%q encoded key>"
"<%q encoded value>"
```

```
bolt page-item --raw-key-only <dbfile> <pageid> <itemid>
<key bytes>
```

```
bolt page-item --raw-value-only <dbfile> <pageid> <itemid>
<value bytes>
```
?

This solves a couple problems:

- It provides a default output that won't mess up someones terminal by outputting raw bytes, and that prints out the key and value, which, if ASCII, will be rendered nicely.
- It avoid use of "\n" as a delimiter between binary output.
- If one needs the raw bytes, they have convenient flags they can use to get the key and value individually.

?

jpbetz commented:
Based on conversation, I've moved things around a bit. Renamed the new subcommand to `page-item` to more clearly reflect it's purpose-- retrieve a page-item.

The default for `page-item` is to print the ascii encoded key and value in separate lines (using go's "%q" []byte formatting).  This default is safe, because it doesn't dump raw bytes to a tty, which can mess up a developers terminal if they don't expect it, and for string keys/values, will print useful output.

To get to raw keys and values, switched to `--raw-key-only`/`--raw-value-only` flags, which should clearly enough hint at their purpose-- print the raw, unencoded, bytes of a key or value.

jpbetz commented:
@heyitsanthony When you get a free minute, would you have one more quick glance at this one?

xiang90 commented on cmd/bolt/main.go:
```diff
@@ -416,9 +419,168 @@ func (cmd *DumpCommand) PrintPage(w io.Writer, r io.ReaderAt, pageID int, pageSi
+	// Parse flags.
+	options := &pageItemOptions{}
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+	fs.BoolVar(&options.keyOnly, "raw-key-only", false, "Directly output the key as raw, unencoded, bytes.")
```
raw or encoded should be another option. i do not like the idea to fix format and output content together.

@jpbetz what do you think?

jpbetz replied:
Yes, that's better, much more compose-able, and we still get our safe, easy to parse defaults. Fixing.

xiang90 commented:
lgtm
